doc/overview.md

```diff
@@ -0,0 +1,97 @@
+1. CircuitDefinition -  definitions contain all the properties of declarations as well as an implementation of the circuit to process the input and produce output.
+    1. Defintions are like function definitions in C, they define how a circuit works in hardware, such as the instances of other circuit definitions used inside of the current circuit.
+    1. Definitions are instances of the DefineCircuitKind class in magma/circuit.py
+1. Instances - these are instances of CircuitDefinitions or CircuitDeclarations. You combine multiple instances to implement another circuit definition.
+
+### How To Create And Test A Circuit In Magma?
+1. Create a module
+1. Implement the module's definition using instances of other modules, like muxes and adders. There are two main ways to do this:
+    1. [Aetherling's upsample](https://github.com/David-Durst/aetherling/blob/master/aetherling/modules/upsample.py#L14) shows how to create a circuit definition using a function that creates an instance of a class which inherits from the Circuit class. The returned class is a CircuitDefinition. Calling the CircuitDefinition creates an instance. When DefineUpsampleParallel() is called, an UpsampleParallel circuit definition is returned. Calling that definition creates an instance. The [UpsampleParallel()](https://github.com/David-Durst/aetherling/blob/master/aetherling/modules/upsample.py#L32) function is a shortcut that creates an UpsampleParallel circuit definition, creates an instance of it, and returns only the instance for use in another circuit definition.
+        1. Note: The canonical approach when creating a circuit defintion is to create both DefineCircuit() and Circuit() functions (like DefineUpsampleParallel() and UpsampleParallel()) where DefineCircuit() returns a circuit definition and Circuit() calls DefineCircuit() and then returns an instance. This makes it easy to use the circuit in other circuits.
+        1. The DefineCircuit() function must always have the @cache_definition annotation to ensure that there is only one instance of each circuit definition with identical parameters. You only want one definition because it is a constructor for producing instances, and there should only be one constructor.
+        1. The class must always have two fields:
+            1. name - The name of the CircuitDefinition in the magma type system
+            1. IO - the input and output ports on the circuit. This is a list that alternates between strings, which are port names, and types, which are the types of the ports. The types should be subclasses of the Type class in magma/t.py.
+    1. [Mantle's counter](https://github.com/phanrahan/mantle/blob/master/mantle/common/counter.py#L25) shows how to create a CircuitDefinition using the DefineCircuit() function call. This is an older circuit that provides more flexibility but is more verbose. It is recommended that users do not use it unless necessary.
```

Comment:
The `DefineCircuit` interface is useful for programmatic construction of circuit definitions. E.g. compiler backends that target magma can use the `DefineCircuit` abstraction to construct and manipulate a magma definition rather than having to construct/manipulate an AST that represents the subclassing `Circuit` syntax.